swarm/network/stream/streamer_test.go

```diff
@@ -921,3 +932,158 @@ func TestMaxPeerServersWithoutUnsubscribe(t *testing.T) {
+			cleanup = func() {
+				os.RemoveAll(datadir)
+				netStore.Close()
+				r.Close()
+			}
+
+			return r, cleanup, nil
+
+		},
+	})
+	defer sim.Close()
+
+	ctx, cancelSimRun := context.WithTimeout(context.Background(), 1*time.Minute)
+	defer cancelSimRun()
+
+	//upload a snapshot
+	err := sim.UploadSnapshot(fmt.Sprintf("testing/snapshot_%d.json", nodeCount))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	//run the simulation
+	result := sim.Run(ctx, func(ctx context.Context, sim *simulation.Simulation) error {
+		log.Info("Simulation running")
+		nodes := sim.Net.Nodes
+		//setup the filter for SubscribeMsg
+		msgs := sim.PeerEvents(
+			context.Background(),
+			sim.NodeIDs(),
+			simulation.NewPeerEventsFilter().ReceivedMessages().Protocol("stream").MsgCode(4), //4 is SubscribeMsg
+		)
+
+		//setup the vars we need
+		msgCount := 0
+		expectedMsgCount := 0
+		allSubscriptionsDone := make(chan struct{})
+
+		//in a loop, catch all SubscribeMsg and just add up
+		go func() {
+			for m := range msgs {
+				if m.Error != nil {
+					log.Error("stream message", "err", m.Error)
+					continue
+				}
+				log.Trace("stream message", "node", m.NodeID, "peer", m.PeerID)
+				//add one
+				msgCount += 1
+				if msgCount == expectedMsgCount {
+					//the expected amount is reached
+					allSubscriptionsDone <- struct{}{}
+					return
+				}
+			}
+		}()
+
+		//first iterate all nodes to get the expected number of subscriptions from the kad table
+		for _, node := range nodes {
+			item, ok := sim.NodeItem(node.ID(), bucketKeyKad)
+			if !ok {
+				return fmt.Errorf("No kademlia")
+			}
+			kad := item.(*network.Kademlia)
+			//define the function which should run for each connection - just count subscriptions
+			//this is not actually subscribing but iterating the same way as the subscriptions do,
+			//as we need just the number
+			eachBinFunc := func(p *network.Peer, bin int) bool {
+				expectedMsgCount += 1
+				return true
+			}
+			//call the actual kademlia for the count
+			kad.EachBin(kad.BaseAddr(), pot.DefaultPof(kad.MaxProxDisplay), 0, eachBinFunc)
+		}
+		log.Debug("Expected message count: ", "expectedMsgCount", expectedMsgCount)
+		//wait until all subscriptions are done
+		<-allSubscriptionsDone
+		log.Info("All subscriptions received")
+		//now iterate again, this time we call each node via RPC to get its subscriptions
+		for _, node := range nodes {
+			//create rpc client
+			client, err := node.Client()
+			if err != nil {
+				t.Fatalf("create node 1 rpc client fail: %v", err)
+			}
+
+			item, ok := sim.NodeItem(node.ID(), bucketKeyRegistry)
+			if !ok {
+				return fmt.Errorf("No registry")
+			}
+			registry := item.(*Registry)
+			//ask it for subscriptions
+			pstreams := make(map[string][]string)
+			err = client.Call(&pstreams, "stream_getPeerSubscriptions")
+			if err != nil {
+				t.Fatal(err)
+			}
+			//length of the subscriptions can not be smaller than number of peers
+			if len(pstreams) < len(registry.peers) {
```

Comment:
Actually, are we absolutely sure here that the message is not only _sent_ but the change also _processed_ at this point, resulting in the change in the streamer state?

Reply:
yes that or message acks after state change :)